### ISSUE TITLE
[Feature] Common Table Expressions (CTE) support

### ISSUE THREAD

**Issue by sdanyliv** [linq2db/linq2db#890]
Originally feature were introduced in #534

For start, here is the great basis article: [SQL Server CTE Basics](https://www.red-gate.com/simple-talk/sql/t-sql-programming/sql-server-cte-basics/)
Also syntax and limitations can be found here: [WITH common_table_expression (Transact-SQL)](https://docs.microsoft.com/en-us/sql/t-sql/queries/with-common-table-expression-transact-sql)
Supported SQL Engines can be found here: [Hierarchical and recursive queries - Common table expression](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#Common_table_expression)

I have prepared proposal syntax how to define CTE in linq2db and it needs your review and suggestions. 

### CTE from IQueryable
It means that IQueryable can be converted to CTE later
```cs 
var cteParent =
(
	from p in db.Parent
	where p.ParentID > 0
	select p
).AsCTE();
```

### IQueryable converted to CTE that uses other CTE
```cs
var cteChildren =
(
	from c in db.Child
	join p in cteParent on c.ParentID equals p.ParentID
	where p.Value1 > 10
	group c by c.ParentID
	into g
	select new
	{
		ParentID = g.Key,
		Count    = g.Count()
	}
).AsCTE();
```

### Query with CTE usage
Nothing special here, just use as IQueryable
```cs
var query = from p in cteParent
	join c in cteChildren on p.ParentID equals c.ParentID
	select new
	{
		p.ParentID,
		p.Value1,
		c.Count
	};
```

### CRUD operations
Also no visual changes
```cs
cteParent.Insert(db.Parent, p => p);

var data = cteParent.ToArray();

cteParent
	.Set(p => p.Value1, p => p.Value1 + 1)
	.Update();

cteParent.Delete();
```

### Updating CTE?
I think we do not need updating CTE it may confuse developers.
```cs 
// this is not new CTE, it's query that uses CTE
var cteParent    = cteParent.Where(p => p.Parent < 100)

// this is new CTE that references previous CTE
var cteParentNew   = cteParent.Where(p => p.Parent < 100).AsCTE()
```

### Recursive CTE
Tricky part, here we need to define class that will be mapped to CTE
```cs
public class CTEClass
{
	public int ChildID;
	public int Count;
}
...

var cteRecursive = db.GetCTE<CTEClass>(cte =>
	from c in db.Child
	join ac in cteChildren on c.ParentID equals ac.ParentID
	join ct in cte on c.ChildID equals ct.ChildID
	select new CTEClass
	{
		ChildID = ct.ChildID + 1,
		Count   = ct.Count + ac.Count
	}
);
```

### New extension methods
```cs
public static class CTEExtensions
{
	public static IQueryable<T> AsCTE<T>(this IQueryable<T> source, string alias = null)
	{
		throw new NotImplementedException();
	}

	public static IQueryable<T> GetCTE<T>(this IDataContext db,
		[InstantHandle] Expression<Func<IQueryable<T>, IQueryable<T>>> cteExpression)
	{
		throw new NotImplementedException();
	}
}
```

### What should be generated 
* `WITH` statement with CTE parts before original SQL. 
* CTE parts will be sorted in order of usage. One CTE can reference only preceding CTEs.
* Circular references will be handled at run-time

**Comment by sergey-miryanov** [comment]
Hi, @sdanyliv will be CTE support implemented?  And if yes when implementation planned?

**Comment by sdanyliv** [comment]
@sergey-miryanov. Will start soon (maybe tomorrow), but i can not estimate how difficult it can be. Will do my best.

**Comment by MaceWindu** [comment]
CTE will be included into 2.0 release (could be tried already from myget)
https://github.com/linq2db/linq2db/wiki/CTE
